instance/models/openedx_instance.py

```diff
@@ -389,3 +394,117 @@ def delete(self, *args, **kwargs):
+            first = first.decode(encoding)
+        if isinstance(second, bytes):
+            second = second.decode(encoding)
+
+        return first == second
+
+    @property
+    def consul_prefix(self):
+        """
+        This is a property that helps determining a specific instance's rpefix in
+        Consul. The prefix helps in keeping different instances' configurations
+        separate from overridings and modifications.
+
+        :return: A unique Key-Value prefix for this instance in Consul.
+        """
+        return '{ocim}/instances/{instance}/'.format(ocim=settings.OCIM_ID, instance=self.pk)
+
+    def _generate_consul_configurations(self):
+        """
+        Collects required configurations for this instance to reflect on Consul.
+        You can add, delete or alter configurations keys and values here which are
+        going to ve reflected in Consul immediately.
+
+        :return: A dict of the configurations.
+        """
+        active_servers = self.get_active_appservers()
+        active_servers_names = [server.name for server in active_servers]
+        basic_auth = self.http_auth_info_base64()
+        enable_health_checks = int(active_servers.count() > 1)
+
+        configurations = {
+            'name': self.name,
+            'domains': json.dumps(self.get_load_balanced_domains()),
+            'health_checks_enabled': str(enable_health_checks),
+            'basic_auth': basic_auth,
+            'active_app_servers': json.dumps(active_servers_names),
+        }
+
+        return configurations
+
+    def _write_configurations_to_consul(self, configurations):
+        """
+        Reflect passed configurations to Consul. Values on consul
+        will be updated only if they changed in this version of the
+        model using agent's `cas` parameter (Check-And-Set).
+
+        If we successfully updated at least one field in Consul
+        then the configurations' version number is gonna be incremented.
+
+        :note: This still doesn't apply removed-configurations case.
+        :param configurations: A dict object contains the configurations
+                               to be written on Consul.
+        :return: A pair (version, changed) with the current version number and
+                 a bool to indicate whether the information was updated.
+        """
+        agent = consul.Consul()
+        contains_updates = False
+
+        version_key = self.consul_prefix + 'version'
+        _, version_data = agent.kv.get(version_key)
+        version_number = int(version_data.get('Value')) if version_data else 0
+
+        for key, value in configurations.items():
+            consul_key = self.consul_prefix + key
+            index, version_data = agent.kv.get(consul_key)
+            cas = index if version_data else 0
+            agent.kv.put(consul_key, value, cas=cas)
+
+            is_updated = not self.compare(value, version_data['Value']) if version_data else True
```

Comment:
I find it hard to understand what exactly is going on here.  I don't like the `OpenEdXInstance.compare()` static method, since I think its name doesn't give any hint as to what it does.  If anything, this should be a function in a utils module.  But prefereably, we don't need this function altogether.  Can't we convert all versions to numbers, and then compare them?

Reply:
I wrote this method to manage the expectations of getting binary/str value back from the agent to allow us comparing mixed types together. I removed this unclear code, and created a helper class that gets and puts the values as expected.